Source/Scene/Scene.js

```diff
@@ -1685,64 +1685,137 @@ define([
         }
-        return [terrainCommands, primitiveCommands];
+    }
+
+    function resetShadowCommands(passCommands) {
+        var length = passCommands.length;
+        for (var i = 0; i < length; ++i) {
+            passCommands[i].length = 0;
+        }
     }
 
     function executeShadowMapCommands(scene) {
         var context = scene.context;
         var uniformState = context.uniformState;
         var shadowMap = scene.shadowMap;
         var isPointLight = shadowMap.isPointLight;
+        var renderState = isPointLight ? shadowMap.pointRenderState : shadowMap.primitiveRenderState;
 
-        var commands = getShadowMapCommands(scene);
-        var terrainCommands = commands[0];
-        var primitiveCommands = commands[1];
+        var shadowPassCommands = shadowMap.passCommands;
+        resetShadowCommands(shadowPassCommands);
 
-        var j;
-        var command;
+        // Insert the scene commands into the shadow map passes
+        var sceneCommands = scene.frameState.commandList;
+        insertShadowCommands(sceneCommands, false, shadowMap, shadowPassCommands);
+        var terrainCommands = getTerrainShadowCommands(scene, shadowMap);
+        insertShadowCommands(terrainCommands, true, shadowMap, shadowPassCommands);
 
         var numberOfPasses = shadowMap.numberOfPasses;
+
+        // // TODO : testing only
+        // for (var k = 0; k < numberOfPasses; ++k) {
+        //     console.log('Pass ' + k + ': ' + shadowPassCommands[k].length + ' commands.');
+        // }
+
         for (var i = 0; i < numberOfPasses; ++i) {
             uniformState.updateCamera(shadowMap.passCameras[i]);
             var passState = shadowMap.passStates[i];
             shadowMap.updatePass(context, i);
 
-            // Execute terrain commands
-            var terrainRenderState = isPointLight ? shadowMap.pointRenderState : shadowMap.terrainRenderState;
-            var numberOfTerrainCommands = terrainCommands.length;
-            for (j = 0; j < numberOfTerrainCommands; ++j) {
-                command = terrainCommands[j];
-                executeCommand(command, scene, context, passState, terrainRenderState, command.shadowCastProgram);
-            }
-
-            // Execute primitive commands
-            var primitiveRenderState = isPointLight ? shadowMap.pointRenderState : shadowMap.primitiveRenderState;
-            var numberOfPrimitiveCommands = primitiveCommands.length;
-            for (j = 0; j < numberOfPrimitiveCommands; ++j) {
-                command = primitiveCommands[j];
-                executeCommand(command, scene, context, passState, primitiveRenderState, command.shadowCastProgram);
+            var passCommands = shadowPassCommands[i];
+            var numberOfCommands = passCommands.length;
+            for (var j = 0; j < numberOfCommands; ++j) {
+                var command = passCommands[j];
+                executeCommand(command, scene, context, passState, renderState, command.shadowCastProgram);
```

Comment:
A bit tangential, but `executeCommand` is becoming unwieldy; we could probably not pass in `context`, and then just access `scene.context` inside it.